esp-hal/src/clock/mod.rs

```diff
@@ -666,3 +670,76 @@ impl<'d> RadioClockController<'d> {
         clocks_ll::reset_rpa();
     }
 }
+
+#[doc(hidden)]
+/// These are all modems with bitmasks as their representation.
+pub enum ModemBitmask {
+    #[cfg(wifi)]
+    WiFi = 1,
+    #[cfg(bt)]
+    Bt = 2,
+    #[cfg(ieee802154)]
+    Ieee802154 = 4
+}
+
+/// Tracks which modems currently request an active PHY clock.
+///
+/// Using an AtomicU8 wrapped in a critical section mutex, isntead of a RefCell, prevents panic
+/// machinery from being generated.
+static PHY_CLOCK_REFS: critical_section::Mutex<AtomicU8> = critical_section::Mutex::new(AtomicU8::new(0));
+
+
+// These functions are moved out of the trait to prevent monomorphization for every modem clock
+// controller. If that doens't really make sense, this can be moved back.
+
+fn enable_phy_clock_internal(modem_bitmask: ModemBitmask) {
+    critical_section::with(|cs| {
+        let inner = PHY_CLOCK_REFS.borrow(cs);
+        let new_phy_clock_ref_count = inner.load(Ordering::Relaxed);
+        if new_phy_clock_ref_count == 0 {
+            clocks_ll::enable_phy(true);
+        }
+        inner.store(new_phy_clock_ref_count | (modem_bitmask as u8), Ordering::Relaxed);
```

Comment:
As long as there is only one place where you can enable a particular clock at a time (i.e. it's not possible to enable wifi clocks twice), this is fine, but is that always ensured?